### PR TITLE
WPB-6577 Fix user creation conflict in SCIM

### PR DESCRIPTION
**Pull request by battermann**
For a description of the manual test approach, see comments of https://wearezeta.atlassian.net/browse/WPB-6577

## Checklist

 - [x] Add a new entry in an appropriate subdirectory of `changelog.d`
 - [x] Read and follow the [PR guidelines](https://docs.wire.com/developer/developer/pr-guidelines.html)

### CODE REVIEW

File: services/spar/src/Spar/Scim/User.hs [hunk 1]
```diff
@@ -750,6 +781,9 @@ deleteScimUser tokeninfo@ScimTokenInfo {stiTeam, stiIdP} uid =
           -- (because that owner won't be managed by SCIM in the first place), but if it ever becomes
           -- possible, we should do a check here and prohibit it.
           unless (userTeam brigUser == Just stiTeam) $
+            -- users from other teams get you a 404.
+            -- users from other teams get you a 404.
+
```
Comment (fisx) [inline services/spar/src/Spar/Scim/User.hs:786]:
```suggestion
```

File: services/spar/src/Spar/Sem/ScimExternalIdStore/Spec.hs [hunk 2]
```diff
@@ -45,15 +46,17 @@ propsForInterpreter interpreter extract lower = do
     prop "insert/lookup" $ prop_insertLookup (Just $ show . void . extract) lower
     prop "insert/insert" $ prop_insertInsert (Just $ show . void . extract) lower
 
+-- FUTUREWORK: Add prop tests for missing operations
```
Comment (fisx) [inline services/spar/src/Spar/Sem/ScimExternalIdStore/Spec.hs:49]: have you given this a try?  if not i can do that.

File: services/spar/src/Spar/Sem/ScimExternalIdStore.hs [hunk 3]
```diff
@@ -22,19 +22,25 @@ module Spar.Sem.ScimExternalIdStore
...
 import Polysemy.Check (deriveGenericK)
+import Spar.Scim.Types
 import Wire.API.User.Identity (Email)
+import Wire.API.User.Scim
 
 data ScimExternalIdStore m a where
   Insert :: TeamId -> Email -> UserId -> ScimExternalIdStore m ()
   Lookup :: TeamId -> Email -> ScimExternalIdStore m (Maybe UserId)
   Delete :: TeamId -> Email -> ScimExternalIdStore m ()
+  InsertStatus :: TeamId -> ValidExternalId -> UserId -> ScimUserCreationStatus -> ScimExternalIdStore m ()
```
Comment (fisx) [inline services/spar/src/Spar/Sem/ScimExternalIdStore.hs:42]:
```suggestion
  -- NB: the fact that we are using `Email` in some cases here and `ValidExternalId` in others has historical reasons (this table was only used for non-saml accounts in the past, now it is used for *all* scim-managed accounts).  the interface would work equally well with just `Text` here (for unvalidated scim external id).
  InsertStatus :: TeamId -> ValidExternalId -> UserId -> ScimUserCreationStatus -> ScimExternalIdStore m ()
```

File: services/spar/src/Spar/Scim/User.hs [hunk 4]
```diff
@@ -530,7 +546,22 @@ createValidScimUser tokeninfo@ScimTokenInfo {stiTeam} vsu@(ST.ValidScimUser veid
...
+      lift $ ScimExternalIdStore.insertStatus stiTeam veid buid ScimUserCreated
       pure storedUser
+  where
+    incompleteUserCreationCleanUp :: UserId -> Scim.ScimError -> Scim.ScimHandler (Sem r) ()
+    incompleteUserCreationCleanUp buid e = do
+      -- something went wrong while storing the user in brig
+      -- we can try clean up now, but if brig is down, we can't do much
+      -- maybe retrying the user creation in brig is also an option?
+      -- after clean up we rethrow the error so the handler returns the correct failure
```
Comment (fisx) [inline services/spar/src/Spar/Scim/User.hs:558]: what about *not* rethrowing, but looking up the status again once or twice?  that'd make for swifter cleanup, especially when the batch job only runs once a day.

**Review by fisx — COMMENTED** [review]
looks good!  i have one actual change request, otherwise looks good!  let me know if you want to discuss this.

File: services/spar/src/Spar/Scim/User.hs (hunk 4 above)
Comment (battermann) [reply services/spar/src/Spar/Scim/User.hs:558]: It only rethrows if `deleteScimUser` fails. AFAICT the deletion is synchronous.

File: services/spar/src/Spar/Sem/ScimExternalIdStore/Spec.hs (hunk 2 above)
Comment (battermann) [reply services/spar/src/Spar/Sem/ScimExternalIdStore/Spec.hs:49]: You can go ahead, but maybe better in a separate PR.

**Review by fisx — APPROVED** [review]

**Pull request merged**